frontend/vite-project/src/types/paymentFormTypes.ts

```diff
@@ -0,0 +1,25 @@
+export interface PaymentFormTypes {
+    cardholderName: string;
+    cardNumber: string;
+    expMonth: string;
+    expYear: string;
+    cvc: string;
+    saveCard: boolean;
+}
+
+export type StringKeys = Exclude<keyof PaymentFormTypes, "saveCard">;
+
+export type ValidationRule = [
+    StringKeys,
+    (value: string) => boolean,
+    string
+];
+
```

Comment:
```suggestion

export interface PaymentActions { 
    setFormData: (data: PaymentFormTypes) => void;
    updateFormField: <K extends keyof PaymentFormTypes>(
        key: K,
        value: PaymentFormTypes[K]
    ) => void;
  }
```
Separate functions that modify state into an actions object.
Combining the actions type with the PaymentState type.